### ISSUE TITLE
Syncing Series is SLOW

### ISSUE THREAD

**Issue by stevezau** [morpheus65535/bazarr#2260]
**Describe the bug**
I've noticed syncing series takes several hours.  It used to be quick..

**To Reproduce**
1. Start the sync series job

**Expected behavior**
Faster than 4-5hrs. 

I have 9000 movies, that syncs fast. I have 3000 series, it takes several hours. 

**Software (please complete the following information):**
Bazarr Version 1.3.0
Package Version v1.3.0-ls218 by linuxserver.io
Sonarr Version 4.0.0.688
Radarr Version 5.0.3.8107
Operating System Linux-6.2.0-33-generic-x86_64-with
Python Version 3.11.5

**Additional context**
What's the best way for me to t-shoot? I can turn on debug but what should i look for?

**Comment by halali** [comment]
we dont support sync all series at all.

but you have to imagine every series have couple episodes... so 3000 series can be like 50000 episodes or so

**Comment by stevezau** [comment]
I mean the "Sync with Sonarr" job. 

It used to take a while on the first run but it was much faster after. But it is still pretty slow. 

I would have thought it would be faster on the following run as the files have no changed?

Maybe ill turn on debug and look at the logs..

**Comment by morpheus65535** [comment]
Should be improved in upcoming beta. I don't have a library that large but let me know how it goes.

**Comment by stevezau** [comment]
Sure, ill test when the new docker image is out. thanks!

**Comment by morpheus65535** [comment]
Can you confirm the improvement? I'm about to merge nightly to main branch and release 1.3.1.

**Comment by stevezau** [comment]
I'm using linuxserver/bazarr:latest atm.

if i switch to the development tag, does that have the changes in ther for me to test?

**Comment by morpheus65535** [comment]
> if i switch to the development tag, does that have the changes in ther for me to test?

Yes

**Comment by stevezau** [comment]
Updated to dev. 

I immediately noticed Movies syncing was MUCH MUCH faster.

However, Series sync still seems slow? How do you want me to t-shoot?

**Comment by morpheus65535** [comment]
Before 1.3.0, we were syncing series and episodes separately so we were getting all episodes metadata from all series, deleting the one that were not present anymore, updating the modified episodes and adding the new one.

Since 1.3.0, series and episodes sync aren't separate process anymore. Now, during the series sync, for each series, we make a call to Sonarr API to list episodes and delete/update/add them to the database. We aren't doing more call to Sonarr, it's all the same. The difference is that we update database after each series instead of one time at the end.

If nothing as changed, it should take long to sync it, maybe a couple of minutes.

I've improved database call to delete/update/insert in batch for a series instead of doing this for each episodes. It should have improve significantly. Didn't you see any difference?

**Comment by stevezau** [comment]
Hmm ok. well it synced 8000 movies in 60 seconds. It has been running for over an hr and a half now and it is only up to series 1029.

Sonarr is minimal CPU. disks at 3% usage. 

So, it does not feel faster to me.

**Comment by stevezau** [comment]
Is there someway i can see where the bottlenecks are?

**Comment by morpheus65535** [comment]
How long did it take before 1.3.0? Anything that I could use as a baseline? I'm going throught the code to see if I can optimize something else.

**Comment by stevezau** [comment]
I am unsue how long it took before but it feels like it is going at the same speed as before. But, as you pointed out it is doing more then before so maybe it is going faster but I'm preceiving it as the same. 

Hmm ok.. I could turn on debugging to see where it is spending most of it's time?

**Comment by morpheus65535** [comment]
> Hmm ok.. I could turn on debugging to see where it is spending most of it's time?

You can but, TBH, I wish you luck trying to analyze the sync log for 3000 series! lol

**Comment by stevezau** [comment]
```13/10/2023 08:58:32|DEBUG   |root                            |BAZARR All episodes from series ID 1883 synced from Sonarr into database.|
13/10/2023 08:58:32|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:32|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"Full Bloom","value":1151,"count":3939}}]|
13/10/2023 08:58:32|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:32|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1889}]|
13/10/2023 08:58:32|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:32|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:32|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1889&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:32|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:32|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1889&apikey=(removed) HTTP/1.1" 200 None|


13/10/2023 08:58:34|DEBUG   |root                            |BAZARR All episodes from series ID 1889 synced from Sonarr into database.|
13/10/2023 08:58:34|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:34|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"The Crimes That Changed Us","value":1152,"count":3939}}]|
13/10/2023 08:58:34|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:34|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1892}]|
13/10/2023 08:58:34|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:34|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:34|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1892&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:34|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:34|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1892&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:35|DEBUG   |root                            |BAZARR All episodes from series ID 1892 synced from Sonarr into database.|
13/10/2023 08:58:35|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:35|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"Moon Knight","value":1153,"count":3939}}]|
13/10/2023 08:58:35|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:35|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1893}]|
13/10/2023 08:58:35|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:35|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:35|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1893&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:35|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:36|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1893&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:36|DEBUG   |root                            |BAZARR All episodes from series ID 1893 synced from Sonarr into database.|
13/10/2023 08:58:36|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:36|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"Hawkeye (2021)","value":1154,"count":3939}}]|
13/10/2023 08:58:36|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:36|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1894}]|
13/10/2023 08:58:36|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:36|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:36|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1894&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:36|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:36|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1894&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:37|DEBUG   |root                            |BAZARR All episodes from series ID 1894 synced from Sonarr into database.|
13/10/2023 08:58:37|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:37|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"Walker","value":1155,"count":3939}}]|
13/10/2023 08:58:37|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:37|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1895}]|
13/10/2023 08:58:37|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:37|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:37|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1895&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:37|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:37|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1895&apikey=(removed) HTTP/1.1" 200 None|



13/10/2023 08:58:44|DEBUG   |root                            |BAZARR All episodes from series ID 1895 synced from Sonarr into database.|
13/10/2023 08:58:44|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:44|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"The Old Man","value":1156,"count":3939}}]|
13/10/2023 08:58:44|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:44|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1896}]|
13/10/2023 08:58:44|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:44|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:44|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1896&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:44|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:44|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1896&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:45|DEBUG   |root                            |BAZARR All episodes from series ID 1896 synced from Sonarr into database.|
13/10/2023 08:58:45|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:45|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"progress","action":"update","payload":{"id":"series_progress","header":"Syncing series...","name":"StartUp","value":1157,"count":3939}}]|
13/10/2023 08:58:45|INFO    |socketio.server                 |emitting event "data" to all [/]|
13/10/2023 08:58:45|INFO    |engineio.server                 |p98NEwXxQiHBMBRgAAAI: Sending packet MESSAGE data 2["data",{"type":"series","action":"update","payload":1897}]|
13/10/2023 08:58:45|DEBUG   |root                            |BAZARR Starting episodes sync from Sonarr.|
13/10/2023 08:58:45|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:45|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episode?seriesId=1897&apikey=(removed) HTTP/1.1" 200 None|
13/10/2023 08:58:45|DEBUG   |urllib3.connectionpool          |Starting new HTTP connection (1): ***.***.***.***:8989|
13/10/2023 08:58:45|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=1897&apikey=(removed) HTTP/1.1" 200 None|```

**Comment by stevezau** [comment]
Hmm, when i watch the logs there is a delay from this line

```
13/10/2023 12:39:45|DEBUG   |urllib3.connectionpool          |http://***.***.***.***:8989 "GET /api/v3/episodeFile?seriesId=5&apikey=(removed) HTTP/1.1" 200 None|```

to the next line.. 

```13/10/2023 12:39:50|DEBUG   |root                            |BAZARR All episodes from series ID 5 synced from Sonarr into database.|```

sometimes it is from 5-30 seconds.

Do you think that's the Sonarr API taking a while to respond? It is not clear if this is an IO wait or if bazarr gets the response and is spending time doing something else.

**Comment by morpheus65535** [comment]
I'm trying different approches to improve time spent by Bazarr to parse the episodes returned by Sonarr but everything I try end up making it worst.

How many episodes does Sonarr have for the series with the longer delay?

**Comment by morpheus65535** [comment]
Determining if values returned by Sonarr API are different from the one stored in database are different take time. I don't see any way to improve this right now.

I've put 4 hours today alone to improve this and haven't found. Let me know if you figure something that I could have miss.

Until then, I guess I'll release 1.3.1 as it is.

Thanks for your involvement!

**Comment by stevezau** [comment]
@morpheus65535 can you point me to the area of code that handles this? i might see if i can live debug what's the bottleneck on my side.

**Comment by morpheus65535** [comment]
For each series, episodes sync is triggered here: https://github.com/morpheus65535/bazarr/blob/master/bazarr/sonarr/sync/series.py#L100

Bottleneck is probably here: https://github.com/morpheus65535/bazarr/blob/master/bazarr/sonarr/sync/episodes.py#L79

**Comment by stevezau** [comment]
@morpheus65535 thanks, im yet to setup a dev env.

I had a quick look at the code tho. does it parse each file, on every run? 

If it does, i wonder how movies are so fast, does it not also parse the movie files?

**Comment by morpheus65535** [comment]
> I had a quick look at the code tho. does it parse each file, on every run?

Do you mean parsing the video files? Unless you've disable the usage of cache in Settings-->Scheduler, it will use the cached ffprobe/mediainfo result to prevent analyzing all the files over and over again.

> If it does, i wonder how movies are so fast, does it not also parse the movie files?

What take time is calling API and parsing the returned JSON. Writing changes to database is also time consuming.

For movies, only one call to Radarr API return all the movies, we parse the JSON once, call the database to compare existing values with parsed JSON returned and do a bulk delete/update/insert to database. Fro each movies, we index external subtitles.

For series, we do one call to get all the series, then, for each series, we store the series to database, call Sonarr again to get the episodes, call it one last time to get the episodes files, call the database to compare existing values with JSON returned and then we bulk delete/update/insert episodes to database. For each episode, we index the external subtitles.

For 3000 series, it means we must call Sonarr 6001 times where it took only one call to Radarr. There's no other way to do it. This is the way Sonarr devs have designed their API and they didn't had any interest in a global episodes API last time I spoke with them.

**Comment by a-priestley** [comment]
> For series, we do one call to get all the series, then, for each series, we store the series to database, call Sonarr again to get the episodes, call it one last time to get the episodes files, call the database to compare existing values with JSON returned and then we bulk delete/update/insert episodes to database. For each episode, we index the external subtitles.

Sorry for the necro-bump, but I am setting up bazarr for the first time and I'm affected by this (I think.. when everything has already synced, sonarr still takes over an hour for me, while radarr takes about a second -- seems like this has yet to be solved).
I understand that the bottleneck is due to the amount of API calls required for sonarr to fetch the required info for every single episode.
I think this could be sped up using ansyncronous I/O.
https://github.com/morpheus65535/bazarr/blob/007c60488c4a2de7d3f324e1213575f2407ceed8/bazarr/sonarr/sync/series.py#L94
Python isn't my strength, but it looks like everything happening inside this loop is done synchronously?

**Comment by morpheus65535** [comment]
> Python isn't my strength, but it looks like everything happening inside this loop is done synchronously?

@a-priestley yes we go one series after the other. You're suggesting we do multiple queries to Sonarr concurrently? Since the delay isn't caused by network latency but by json serialization and other CPU bound, we would possibly hit another wall here. On top of that, it increase complexity of progress reporting in webui and debugging in logs that would get mixed up.

**Comment by a-priestley** [comment]
@morpheus65535 I suggest using concurrency as a general approach, but I wouldn't mind testing some things to confirm if/where/how it would be beneficial. With the added concerns about progress tracking and debugging, I can certainly *attempt* to improve things and let you know how it goes! (again python is not something I normally write in, but I think my fundamentals are good enough)

**Comment by morpheus65535** [comment]
@a-priestley let me know a bit more about your strategy before commiting too much time to a PR. You can reach the dev teams and other contributors on the dev-community channel on our Discord server.

**Comment by nodiaque** [comment]
Was this issue closed? On my bazarr instance, I don't know when it started but it used to run fine. But now, it's just running foerever. I just open my sonarr instance, it had 4 job running a sonarr sync (a check should be made to not schedule another sync when it's already sync) and I had 51 jobs queue, 40 were sonarr sync. Because of that, Bazarr was basicly doing nothing. I switched to a 6 hours sync with sonarr, which mean it doesn't detect new series as fast as it used to, hoping it improve.

**Comment by morpheus65535** [comment]
@nodiaque it's fixed in upcoming 1.6.0
